### ISSUE TITLE
Add scalaz.effect Type Class Hierarchy

### ISSUE THREAD

**Issue by jdegoes** [scalaz/scalaz#1725]
Working notes.

```scala
sealed trait ExitResult[E, A]
case class Terminated[E, A](cause: Throwable) extends ExitResult[E, A]
case class Failed[E, A](error: E) extends ExitResult[E, A]
case class Completed[E, A](value: A) extends ExitResult[E, A]

sealed trait Fiber[F[_, _], E, A] {
  def join: F[E, A]
  
  def interrupt[E2](t: Throwable): F[E2, Unit]
}

trait FallibleClass[F[_, _]] extends BifunctorClass[F] {
  def applicativeClass[E]: ApplicativeClass[F[E, ?]]
  
  def attempt[E1, E2, A](b: F[E1, A]): F[E2, E1 \/ A]
  
  def fail[E, A](e: E): F[E, A]
}

trait TerminatableClass[F[_, _]] extends FallibleClass[F] {
  def monadClass[E]: MonadClass[F[E, ?]]

  def terminate[E, A](t: Throwable): F[E, A]
  
  def always[E, A](b: F[E, A])(f: ExitResult[E, A] => F[Void, Unit]): F[E, A]
  
  def execute[E1, E2, A](b: F[E1, A]): F[E2, ExitResult[E1, A]]
}

trait ForeignClass[F[_, _]] extends TerminatableClass[F] {
  def sync[E, A](effect: => A): F[E, A]
  
  def async[E, A](register: (ExitResult[E, A] => Unit) => Unit): F[E, A]

  def asyncPure[E, A](register: (ExitResult[E, A] => F[Void, Unit]) => F[E, Unit]): F[E, A]
}

/**
 * `race(io, never) == race(never, io) == io`
 * `fork >=> join == identity`
 * `fork never >=> interrupt t == point ()`
 */
trait ConcurrentClass[F[_, _]] extends TerminatableClass[F] {
  def never[E, A]: F[E, A]

  def uninterruptible[E, A](f: F[E, A]): F[E, A]

  def fork[E2, E, A](fa: F[E, A]): F[E2, Fiber[F, E, A]]
  
  def raceWith[E, A, F, C](l: F[E, A], r: F[E, F])(finish: (A, Fiber[F, E, F]) \/ (F, Fiber[F, E, A]) => F[E, C]): F[E, C]
}

trait EffectClass[F[_, _]] extends ForeignClass[F] with ConcurrentClass[F]
```

**Comment by durban** [comment]
> `fork f >>= \f -> interrupt f t *> join f == terminate t`

Is this really true? If I understand it correctly, the left hand side of the `==` concurrently starts `f`, then immediately interrupts and then joins it. But if the forked `f` is "fast enough", it can already finish before the interruption happens. Am I missing something?

**Comment by jdegoes** [comment]
The law is wrong for 2 reasons: the reason you state, and another reason, which is that it may execute some part of `f`, which makes it not equivalent to `terminate t`. 

I do want to see if I can turn it into a law but it will not be so simple as written. 😄 

Help welcome on this or other type class laws. 😄

**Comment by jdegoes** [comment]
I will add a note here to extract a law from this:

```
fork f >>= \f -> interrupt f t *> join f == terminate t
```

**Comment by durban** [comment]
I'm trying to understand this hierarchy, so I'm trying to figure out where bracket fits in. And I have trouble writing it with these operations. What I have so far is this:
```scala
// TODO: this seems incorrect
def bracket[E, A, B](acq: F[E, A])(use: A => F[E, B])(rel: A => F[Void, Unit]): F[E, B] = {
  monadClass.flatMap(uninterruptible(acq)) { a =>
    always(use(a)) { _ =>
      rel(a)
    }
  }
}
```

But this doesn't seem correct, since if it gets interrupted immediately after `uninterruptible(acq)` succeeds, then we don't even enter the `always ...` part, so `rel` will not run ...

**Comment by jdegoes** [comment]
@durban My first goal is to eliminate `bracket` as a primitive. If I can't, then it will replace `always`. If I can, it will go alongside `always` so it can be optimized specially.

I think you can eliminate `bracket` as a primitive using `Promise` and `fork`, by having complex coordination in which you `always` the fork (but not the `use` being forked), run the `use` separately, and then in the main fiber (which is uninterruptible), it will `release` as soon as the forked fiber terminates. If this can be sketched out, it means `bracket` is not primitive and can appear merely as an optimization.

**Comment by durban** [comment]
@jdegoes Yes, it may be possible with `Promise`. But then the next question is: is `Promise` a primitive? (If not, then implementing it would surely need what's called above `ForeignClass`; that would mean bracket cannot be on `TerminatableClass`.)

**Comment by jdegoes** [comment]
@durban I found it only needed `IORef` and `ensuring` (e.g. `always` above). You can see the implementation [here](http://github.com/scalaz/scalaz-zio).

I still don't like `async` and `asyncPure` and `sync`.

**Comment by jdegoes** [comment]
Maybe `IORef` should be treated as primitive. It's the `IO` version of a mutable reference—hard to build much imperative code without it. You could create them from a type class.

`ensuring` is weaker than `always` above, which might affect the implementation. But you still need a ref to do `bracket`. We could go back to making `bracket` primitive or think if there's something(s) more primitive than `bracket` but more powerful than `always`.
